The return of the GroupBy and GroupByMut iterators on slice

Kerollmops opened the pull request:
According to https://github.com/rust-lang/rfcs/pull/2477#issuecomment-742034372, I am opening this PR again, this time I implemented it in safe Rust only, it is therefore much easier to read and is completely safe.

This PR proposes to add two new methods to the slice, the `group_by` and `group_by_mut`. These two methods provide a way to iterate over non-overlapping sub-slices of a base slice that are separated by the predicate given by the user (e.g. `Partial::eq`, `|a, b| a.abs() < b.abs()`).

```rust
let slice = &[1, 1, 1, 3, 3, 2, 2, 2];

let mut iter = slice.group_by(|a, b| a == b);
assert_eq!(iter.next(), Some(&[1, 1, 1][..]));
assert_eq!(iter.next(), Some(&[3, 3][..]));
assert_eq!(iter.next(), Some(&[2, 2, 2][..]));
assert_eq!(iter.next(), None);
```

[An RFC](https://github.com/rust-lang/rfcs/pull/2477) was open 2 years ago but wasn't necessary.

rylev commented on library/alloc/tests/slice.rs:
```diff
@@ -1898,3 +1898,49 @@ fn subslice_patterns() {
 }
+
+#[test]
+fn test_group_by() {
```
It would be nice to have elements that are equal to previous elements but are disjoint from their group. For example:

```rust
let slice = &[1, 1, 1, 3, 3, 2, 2, 2, 1];
```

Kerollmops commented:
Hey @rylev,

Thanks for the comment, I changed the tests a little bit, however, do you have an idea on how I should fix [this CI tidy issue](https://github.com/rust-lang/rust/pull/79895/checks?check_run_id=1530096463#step:24:1942)? The file is now too long what should I do? Create a new one just for those two structs?

rylev commented:
@Kerollmops I'm not sure. I'd also be curious to know what the right move is here (either to suppress the warning or to break up the file).

lcnr commented:
I think it's fine to just suppress the warning for now. Splitting `slice::iter` into submodules is something that's easier in a followup PR

pickfire commented on library/core/src/slice/iter.rs:
```diff
@@ -2967,3 +2968,167 @@ unsafe impl<'a, T> TrustedRandomAccess for IterMut<'a, T> {
+/// [`group_by`]: ../../std/primitive.slice.html#method.group_by
+/// [slices]: ../../std/primitive.slice.html
+#[unstable(feature = "slice_group_by", issue = "none")]
+#[derive(Debug)] // FIXME implement Debug to be more user friendly
```
Will this get fixed after pull request or as part of the pull request?

Kerollmops replied:
I marked it there, I will probably fix it in this PR, I forgot about that, thanks for reminding me.

lcnr commented:
cc @rust-lang/infra bors didn't assign anyone here

r? @m-ou-se

camelid commented:
> bors didn't assign anyone here

FYI I think you mean `rust-highfive`. Probably the bot was being redeployed or something.

m-ou-se reviewed (commented):
Thanks! This looks great!

> 2 years ago

Apologies again! We're working on streamlining our processes. This shouldn't happen again. :)

> Create a new one just for those two structs?

We should probably split files like these into more submodules at some point (just like #78934 does). A separate file for these new structs would be good (e.g. `slice/iter/group_by.rs`). But leaving that for later is also fine. Then it can be done together with the rest of the file.

---

Only one thing left before we can merge this:

Can you [open a tracking issue](https://github.com/rust-lang/rust/issues/new?labels=C-tracking-issue%2C+T-libs&template=library_tracking_issue.md&title=Tracking+Issue+for+XXX) for this, and add its number in the `unstable` attributes? Thanks!


(By the way, feel free to rebase/squash commits.)

m-ou-se commented on library/core/src/slice/mod.rs:
````diff
@@ -1207,6 +1210,66 @@ impl<T> [T] {
+    /// assert_eq!(iter.next(), Some(&[3, 3][..]));
+    /// assert_eq!(iter.next(), Some(&[2, 2, 2][..]));
+    /// assert_eq!(iter.next(), None);
+    /// ```
````
Just a suggestion: It wasn't obvious to me at first that this function could be used with e.g. `a < b` to get the sorted subslices. (e.g. `[1,2,3,2,3]` -> `[&[1,2,3], &[2,3]]`) Maybe an extra example like that here could show this flexibility of `group_by`.

Kerollmops replied:
Thank you, I have added an extra example for that.

Kerollmops commented:
I just created [the tracking issue](https://github.com/rust-lang/rust/issues/80552), added it to the files, and squashed everything.

No worries about the 2 years waiting time, [I implemented it myself, it was quite fun](https://github.com/Kerollmops/slice-group-by)! I also benchmarked it and the full safe Rust version [is a little bit slower than the unsafe one](https://github.com/Kerollmops/slice-group-by/pull/19).

m-ou-se commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 8b53be660444d736bb6a6e1c6ba42c8180c968e7 has been approved by `m-ou-se`

<!-- @bors r=m-ou-se 8b53be660444d736bb6a6e1c6ba42c8180c968e7 -->
<!-- homu: {"type":"Approved","sha":"8b53be660444d736bb6a6e1c6ba42c8180c968e7","approver":"m-ou-se"} -->

SkiFire13 commented on library/core/src/slice/mod.rs:
```diff
@@ -1207,6 +1210,96 @@ impl<T> [T] {
         RChunksExactMut::new(self, chunk_size)
     }
 
+    /// Returns an iterator over the slice producing non-overlapping runs
+    /// of elements using the predicate to separate them.
```
I feel like this should have a stronger mark on the fact that it only groups **consecutive** items, especially with such an ambiguous function name.

m-ou-se replied:
I'd say 'runs of elements' already implies that. But I'd be happy to review your PR if you have an idea on how to improve the wording. Improvements to documentation are always welcome. :)

SkiFire13 commented on library/core/src/slice/mod.rs:
````diff
@@ -1207,6 +1210,96 @@ impl<T> [T] {
+    /// ```
+    /// #![feature(slice_group_by)]
+    ///
+    /// let slice = &[1, 1, 1, 3, 3, 2, 2, 2];
+    ///
+    /// let mut iter = slice.group_by(|a, b| a == b);
+    ///
+    /// assert_eq!(iter.next(), Some(&[1, 1, 1][..]));
+    /// assert_eq!(iter.next(), Some(&[3, 3][..]));
+    /// assert_eq!(iter.next(), Some(&[2, 2, 2][..]));
+    /// assert_eq!(iter.next(), None);
````
I feel like this gives a false perception that it groups the same items, no matter where they are. Would probably be better if the same element appeared multiple times in different groups, for example:

```suggestion
    /// let slice = &[1, 1, 1, 3, 3, 2, 2, 2, 1, 1];
    ///
    /// let mut iter = slice.group_by(|a, b| a == b);
    ///
    /// assert_eq!(iter.next(), Some(&[1, 1, 1][..]));
    /// assert_eq!(iter.next(), Some(&[3, 3][..]));
    /// assert_eq!(iter.next(), Some(&[2, 2, 2][..]));
    /// assert_eq!(iter.next(), Some(&[1, 1][..]));
    /// assert_eq!(iter.next(), None);
```

Kerollmops replied:
I understand your point but I find the second example clear enough and let the user understand that the split are based on the specified function.

SkiFire13 replied:
What if someone skips the second example? You could argue that's the user's fault, but I'd like to make this as much foolproof as we can

bors commented:
:hourglass: Testing commit 8b53be660444d736bb6a6e1c6ba42c8180c968e7 with merge b33e234155b33ab6bce280fb2445b62b68622b61...
<!-- homu: {"type":"BuildStarted","head_sha":"8b53be660444d736bb6a6e1c6ba42c8180c968e7","merge_sha":"b33e234155b33ab6bce280fb2445b62b68622b61"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1630631556)
Approved by: m-ou-se
Pushing b33e234155b33ab6bce280fb2445b62b68622b61 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"m-ou-se","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1630631556"},"merge_sha":"b33e234155b33ab6bce280fb2445b62b68622b61"} -->

Kerollmops commented:
Thank you for the review @m-ou-se :)
